### ISSUE TITLE
templates vs types

### ISSUE THREAD

**Issue by fosterlynn** [valueflows/valueflows#185]
@elf-pavlik and I had a long chat in gitter the other day that I want to bring into focus here.  It involves whether the things we currently call XTemplate, in the top layer are "templates" or "types".  This was after the also extended discussion of template/type/model in the [this PR](https://github.com/valueflows/valueflows/pull/176) and in #179 .

In the PR, I suggested we call those things in the top layer Template, because I thought we had agreed that although they are "type objects", the name Type is confusing because of the use of "type" as in rdf:type for a LOD class/instance kind of relationship.

Then in the gitter chat I found out that Pavlik thought template != type, while I thought template == type. I think we are both representing things the vocabulary needs, we just need to figure out what to call them, and where they fit in the model, and whether they go in the "top" layer or as "plan" layer, which aren't part of the vocab strictly, but do give good conceptual structure I think.  I think it involves whether the Templates can be used with some things "filled in" already - for example agents and quantities in an exchange template, so they can be used like an agreement for things they want to commit to later. 

Thus, this issue to figure it out.  There are a bunch of subtleties involved here.

**Comment by ahdinosaur** [comment]
can see why Type is overloaded, since it has many meanings across programming paradigms that might be confusing. can also see why Template doesn't exactly convey the correct meaning.

other potential words for the soup: abstract, blueprint, prototype, concept, scheme, formula?

**Comment by bhaugen** [comment]
I'm doing a bunch of research to differentiate type, template, class, category, and taxonomy, but it's pretty squishy and will take me at least until to tomorrow to disambiguate.

But as a preliminary, I think we need to remember that our domain is economic relationships, and not programming languages (although the programming language usage is informative).

So, for example, I think type in economic relationships tells you what something is. As in exchanges, if somebody gives you a Grade A Organic Haralson Apple, you pretty much know what that is.

On the other hand, a Template as (for example) an element in a recipe is used to specify (or as a factor in computing) a lot of the attributes of an input or output. For example, the resource type (which, in the context of this conversation is a Type, not a Template), and the quantity (which is not meant to be used as-is, which is a little un-template-like, but extended by the quantity to be produced: for example, if 2 Apples specified in the Applesauce input recipe element, and 10 units of Applesauce are to be produced, then 20  Apple inputs are required).

I'll continue with the research. This is a preliminary opinion.

**Comment by fosterlynn** [comment]
hi @ahdinosaur really good to see you pop in over here :heart:

**Comment by elf-pavlik** [comment]
> In the PR, I suggested we call those things in the top layer Template, because I thought we had agreed that although they are "type objects", the name Type is confusing because of the use of "type" as in rdf:type for a LOD class/instance kind of relationship.

I think one of the differences between how I see Template compared to 'Type Object', comes with preference to 'stamp' the instance using a 'template' and copy over values directly on the instance. While 'type object' would just have instance referencing it but not have all the values directly (as copies).

> Templates can be used with some things "filled in" already - for example agents and quantities in an exchange template, so they can be used like an agreement for things they want to commit to later.

When we 'stamp' instances with a template w can have some references already preset, what I see as benefit over 'type objects' - if some references on Template change, instances will keep their own references which their copied from Template during stamping it or have changed later.

> So, for example, I think type in economic relationships tells you what something is. As in exchanges, if somebody gives you a Grade A Organic Haralson Apple, you pretty much know what that is.

I think currently categories serve the purpose of answering 'belongs to or doesn't belong' (which I find clearer then ambiguous *is*).
When it comes to the discussion #179 Model/Template, If we take as an example Fairphone2 which has weight: 148g, currently I see benefits of having copy of that value directly on each instance of Fairphone2 model, so that queries don't need to check if instance references some 'type object' and look for values there. Again this relates to data availability and need for caching data referenced as 'type object' and if those 'type objects' change some values keeping them in sync and making sure that some undesired change in 'type object' doesn't get reflected on resource referencing it.

**Comment by pmackay** [comment]
> the name Type is confusing because of the use of "type" as in rdf:type for a LOD class/instance kind of relationship.

I dont personally find a problem with this. If we're writing RDF/TTL files and defining rdf:type its pretty clear IMHO. In other more document-y contexts should the convention be to say "RDF type" or similar to be clear on a distinction? But then any ObjectTypes would have distinct names anyway.

>I think one of the differences between how I see Template compared to 'Type Object', comes with preference to 'stamp' the instance using a 'template' and copy over values directly on the instance. 

Is the detail of whether values are copied into an instance object something that should inform the naming of the vocab? We could use "Type Object" naming and still any implementation could do this copying approach. But to prefer Template to imply that will always happen doesnt seem quite right IMHO.

**Comment by fosterlynn** [comment]
>When it comes to the discussion #179 Model/Template, If we take as an example Fairphone2 which has weight: 148g, currently I see benefits of having copy of that value directly on each instance of Fairphone2 model, so that queries don't need to check if instance references some 'type object' and look for values there. Again this relates to data availability and need for caching data referenced as 'type object' and if those 'type objects' change some values keeping them in sync and making sure that some undesired change in 'type object' doesn't get reflected on resource referencing it (different agents can have control over 'type object' record and resource record!)

@elf-pavlik do you think we should also have type objects?  Seems like in this example, even if you pull data down from the type object to the plan or observation layer object (which you point out potentially good reasons for), you still need the type object in the first place, true?

>Is the detail of whether values are copied into an instance object something that should inform the naming of the vocab? We could use "Type Object" naming and still any implementation could do this copying approach. But to prefer Template to imply that will always happen doesnt seem quite right IMHO.

I agree, I think we are leading to two different things.  I think we need type objects - whatever we call them, and I'm also fine with naming them XType and being clear in the doc, as suggested by @pmackay .  If there are no type objects, template objects will proliferate because they will be agent specific and it will be difficult to figure out anything like the type.

I'm fine with having something that isn't type objects that can be pre-filled so to speak, that is more like a template.  I think those would live on the Plan layer, and probably be some kind of an agreement.  If the name XTemplate fits there, then fine.

Recipe things, especially what I called EventTemplate, have a little flavor of a template because they have quantities.  But I still think they are a type, you can't just take it and fill it in, because when creating a plan you have to calculate the planned quantities from how many you want, and multiply appropriately the quantities in the recipe.

**Comment by elf-pavlik** [comment]
> @elf-pavlik do you think we should also have type objects? Seems like in this example, even if you pull data down from the type object to the plan or observation layer object (which you point out potentially good reasons for), you still need the type object in the first place, true?

I think this 'Type Object' relates pretty much to 'Resource Model' which I think we may better discuss as specific case in #179 Looking at your [latest UML](https://raw.githubusercontent.com/valueflows/valueflows/405ab262e4b6aa88604c4b16bb3da8b791265434/release-doc-in-process/ValueFlowsUML.png) other *Template doesn't seem 'Type Objects' since they only suggest *name* and they name of the template probably will not get copied to the instance based on the template?
I'll make PR to replace *vf:RelationshipTemplate* with *rdf:Property* and mark *vf:Action* as *rdf:subClassOf* *rdf:Property*. This will leave just
* ProcessTemplate
* ExchangeTemplate
* EventTemplate

Which don't seem to work as 'TypeObject', but maybe we need some examples of them in YAML to clearly discuss them. Also as a reminder, we discussed Templates as something just used in Recipe, one can do all the planning without ever referencing any Template as long as one doesn't try to create Recipe. Which again makes them seem to serve different purpose than 'TypeObjects'.

When it comes to EventTemplate, we discussed Intent & Commitment as 'event image' - which in a way makes them kind of a 'Template' for events that will fulfill them!

**Comment by bhaugen** [comment]
Some results from research, which is really varied and often discordant (lots of disagreements), but the consensus seems to be:
* a Type is a formalization of a concept. An abstraction. In software, a Type is about behavior, not implementation.
* whereas a Class, or Template, is about implementation. For a Class, how a Type is implemented in software. A Template is a more general concept, but also has to do with how something is created. Dictionary definition: http://www.dictionary.com/browse/template
> a pattern, mold, or the like, usually consisting of a thin plate of wood or metal, serving as a gauge or guide in mechanical work.

**Comment by fosterlynn** [comment]
>I'll make PR to replace vf:RelationshipTemplate with rdf:Property and mark vf:Action as rdf:subClassOf rdf:Property.

Actually I did that on purpose.  I've been thinking it makes more sense for the UML version to support more the non-LOD crowd.  The VOWL version should show the LOD version, and it should probably go up top.  But for example, @ahdinosaur is working on a project using VF that won't be LOD. And if people just want to see what it might look like in an object model, or from a logical perspective, "Property" isn't helpful.

I do need to review the VOWL version (thanks @djodjoni ! ), and maybe that is even what you meant?

**Comment by fosterlynn** [comment]
>Also as a reminder, we discussed Templates as something just used in Recipe, one can do all the planning without ever referencing any Template as long as one doesn't try to create Recipe. 

I don't understand this one.  How would templates as you are defining them be used in recipes? 

And do you mean people can plan without recipes?  I agree with that.  Or something else?

**Comment by djodjoni** [comment]
A little late input for what it's worth. IMO although Types are truly overloaded they are less confusing than Templates.
Also as @bhaugen  mentioned 
> But as a preliminary, I think we need to remember that our domain is economic relationships, and not programming languages (although the programming language usage is informative).

So the RDF or any programming language type here should not be relevant. 

As regards to copying params to instances then it means that the instances should have at least the same properties which mean that is kind of 'subclass' which means that type as 'meta type' is actually not that bad (infact actual software implementation may used it exactly like this). Anyway template or type for me it doesn't matter too much but explanation and examples should be there.

Btw mentioning this are there already some discussions/agreements about the rules of copying/caching params/refs to the instance ?

**Comment by ahdinosaur** [comment]
yeah, as @fosterlynn mentioned, i'm starting to write the code for [cobuy](https://github.com/enspiral-root-systems/cobuy), where the plan is to use Value Flows data structures but with a ["kappa architecture"](https://www.confluent.io/blog/turning-the-database-inside-out-with-apache-samza/) database similar to [Scuttlebutt](https://www.scuttlebutt.nz/). will try and join the conversations again more, apologies for my prolonged absence! am appreciating the progress that's happened while i've been away. :heart: hopefully soon i'll have some code worth showing and will start a targeted discussion about applying a subset of the vocabulary to our particular use case (food buying groups) and database (source of truth is an append-only log of immutable messages).

**Comment by fosterlynn** [comment]
I'd love to move towards making a decision on this one if we can.  Proposal for discussion:  

1. Rename the classes in the "top/abstract/knowledge" layer to:
* ResourceType
* ProcessType
* IntentTemplate (might need to change this one to something not "type" and not "template", but for now.....)
* Action (keep it)
* ExchangeType (to be defined in detail)
* TransferType (if we have those - I need to rethink the type layer for exchange in our new model - but I think we will still have them, and if so they are a type object - won't include any agents, probably no quantities, will be a way to define how the network exchanges in general)

2. Create some new classes for what @elf-pavlik wants to do with templates, as partially pre-filled things that you can use to more easily plan for repeating exchanges or processes.  These would be on the "plan" layer.  And/or whatever else Pavlik thinks is important to have in the way of Templates.

And, here is a yaml snippet for those inclined.  This is so Pavlik can say what else he would have in a template than what is here.  Disclaimer: I don't know exactly how to do the vf:resource / vf:category properties, so please feel free to correct me as needed.  And of course the same on anything else, I expect I will need it. :)

``` yaml
'@context': https://w3id.org/valueflows/v1-dev
'@graph':
  - '@id': ex:cook-salsa
    '@type':
      - vf:ProcessType
    'vf:estimatedDuration': 
        'vf:quantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Hour
          'quid:numericValue': 5
    'vf:note': 'Combine all ingredients, ....'
    'vf:input':
      - '@id': '#recipe-input-1'
        '@type': vf:IntentTemplate
        'vf:action': vf:consume
        'vf:category': prodont:Tomato
        'vf:affectedQuantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Kilogram
          'quid:numericValue': 4
      - '@id': '#recipe-input-2'
        '@type': vf:IntentTemplate
        'vf:action': vf:consume
        'vf:category': prodont:Onion
        'vf:quantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Kilogram
          'quid:numericValue': 2
      - '@id': '#recipe-input-4'
        '@type': vf:IntentTemplate
        'vf:action': vf:work
        'vf:category': ex:Cooking
        'vf:quantity':
            '@type': quid:QuantityValue
            'quid:unit': quid:Hour
            'quid:numericValue': 3
    'vf:output':
      - '@id': '#recipe-output-1'
        '@type': vf:IntentTemplate
        'vf:action': vf:produce
        'vf:resource':
          '@type': vf:ResourceType
          '@id': www.example.com/products/ex:Salsa-.25l-12331213
        'vf:quantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Each #or however we do this
          'quid:numericValue': 12

**Comment by fosterlynn** [comment]
From gitter chat, I think @elf-pavlik and I have agreed to separate the 2 (different) things we thought were on the former "recipe" layer, the one on the top.  It is something like types and templates.  

Any objections?

Templates might get their own layer.  Top layer name ideas: "abstract", "knowledge".  (Noting the layers are just conceptual, to help us and others understand, they are not actually represented in the vocabulary.)

**Comment by pmackay** [comment]
@fosterlynn I think your point 1 above sounds good. There are aspects of the YAML snippet I probably need to spend more time reviewing.

Just a thought, might it be worth expressing in the introduction parts of the docs high level descriptions of the layers (recipe, plan, etc) and how/where they get used? Could be great for setting context :)

**Comment by bhaugen** [comment]
Is your definition something like?
* A type declares what something is, to differentiate from other types of the same row and column, for example, a bike is not an apple, and a Delicious apple is not a Haralsom apple.
* A template is used in the construction of a record, for example, to set default attribute values.

Feel free to wordsmith better. But I think we should publish definitions.

**Comment by fosterlynn** [comment]
>Just a thought, might it be worth expressing in the introduction parts of the docs high level descriptions of the layers (recipe, plan, etc) and how/where they get used? Could be great for setting context :)

:+1:   I can work on that at some point..... probably not next though. :blush:

**Comment by bhaugen** [comment]
This might be useful, with some tweaks. I used it in a discussion of refactoring our NRP software.
https://github.com/django-rea/nrp/issues/1

Columns and rows.

Layer | Resource | Event | Agent | Process | Exchange
--------|--------------|---------|----------|------------|---------------
Recipe/Type | ResourceType | EventType, CommitmentType | AgentType | ProcessType | ExchangeType
Plan | | Commitment | | Process | Exchange
Reality | Resource | EconomicEvent | Agent | Process | Exchange

**Comment by fosterlynn** [comment]
:+1:  to getting some definitions clearly stated.  @bhaugen 's is pretty clear, except "row and column" is muddy for me.

>  * A type declares what something is, to differentiate from other types of the same row and column, for example, a bike is not an apple, and a Delicious apple is not a Haralsom apple.
>  * A template is used in the construction of a record, for example, to set default attribute values.

Our category acts like a type object I think, just generalized.  I think we may have to allow categories as resource types in recipes, hmmm....  And I think for some, like Action, there is definite behavior implied.  

In the case of VF, I think our type objects should not have any agents. Whether general or very specific, they should be referable (re-usable) by anyone who wants to.  In some other properties besides agent, there may be some gray area, like I'm fine not calling the thing currently called IntentTemplate a type.  But for the template side, it can have agents, and in general have more stuff filled in already so it is very practical for creating recurring intents or commitments.

I need to define exchange types to fit our new exchange model, and then see what that looks like.  And see if I can fit them into recipes.  Then get more serious about those names and definitions.

**Comment by djodjoni** [comment]
nice example ! however about the vf:action isnt it supposed to be part of the process, not the intent ?

I am still not sure how we all see types/templates :) as in practice they can be seen as a single concept, i.e. a type(as it defines what smth is) can clearly define its default values and Template as it defines how something is created can be used to identify what that thing is, but this goes to more philosophical discussion :) I suppose well see how things works in practice or with more examples.

btw I am not also sure how vf:action could be useful the way it is in the example. If I recursively fall in abstractions I could see it as nano process. Which also means it has tho have an io as well.

Right now I see it as 'type/cat like thing':
'vf:action': vf:consume -> consumable/ingredient
'vf:action': vf:work -> service/process

which can be deduced by the category or type.

**Comment by bhaugen** [comment]
@djodjoni 
> I am still not sure how we all see types/templates :) as in practice they can be seen as a single concept, 

That's what they were, and I am still happy with that name. Then they got renamed and a lot of dithering has gone on, and now we have both names, and I want us to define what we mean by them. Or go back to one name.

Action used to be EventType in NRP. It encapsulates the resource effect (increment, decrement, none, etc) and whether it applies to a process or an exchange (or maybe both?) Otherwise those behaviors get duplicated all over Events, Commitments, Recipes, etc.

**Comment by fosterlynn** [comment]
@djodjoni As you can tell, we have ongoing questions around translating some of the things that started as types in OO land (actions, relationship types for example) to LOD in a way LOD people will naturally think of them.  In the process we have lost behavior related properties of some of them.  In the case of vf:action, we accommodated the loss of resource effect by inheritance of the properties, which you will see in the latest ttl.

But in any case, they do apply to events (and event-like things: commitments, intents (we think), and the event piece in the recipe (IntentTemplate or whatever it ends up being)).  The events define all the motion ("action") in the REA model - what happens to resources when they are input to or output from a process.

>Right now I see it as 'type/cat like thing':
>'vf:action': vf:consume -> consumable/ingredient
>'vf:action': vf:work -> service/process

Yes, I agree it is a type/cat thing.  But it is more like vf:consume describes what the event represents, not the resource itself - yes it is consumable, but now it actually got consumed; and vf:work describes that this event means some work did happen.  The event also will reference the resource that it affects through this action.

**Comment by elf-pavlik** [comment]
> I'll make PR to replace vf:RelationshipTemplate with rdf:Property and mark vf:Action as rdf:subClassOf rdf:Property. This will leave just
>*  ProcessTemplate
>*  ExchangeTemplate
>*  EventTemplate

I would like ti stay careful not to create to much confusion around 'Types' and 'TypeObjects'. We pretty much already clarified that what UML called EventTemplate and we just renamed to IntentTemplate **does NOT** act as anything like 'TypeObject'. Instances of Intent (possibly also directly Commitment or Event) created based on IntentTemplate don't have to reference that tempate at all, so **NOT** a type object.

Still keeping 'ResourceMode' aside, we have proposed
* ExchangeTemplate
* ProcessTemplate

I think they need some clear examples. In @fosterlynn example above the
```yaml
 - '@id': ex:cook-salsa
    '@type':
      - vf:ProcessType
    'vf:estimatedDuration': 
        'vf:quantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Hour
          'quid:numericValue': 5
```

also **does NOT** seem to act as 'TypeObject', instance of *vf:Process* created based on this template, doesn't have to reference it at all. I guess we will have exactly the same situation for proposed 'vf:ExchangeTemplate'.

What we currently have as *vf:Action* defined as *rdfs:subClassOf rdf:Property*, may have some notion for 'TypeObject' if we rely on it to know if Event should for example decrement or increment resource which it affects. PR #192 shows that Event could do that without relying on 'action verb'.

We also have *vf:category*, which we may also want to use for categorizing processes. As I explained before, it works purely to categorize something using some taxonomy graph (hierarchy which allows multiple parents). I still wouldn't call those categories 'TypeObjects' since the instance which references that category doesn't rely on any of the properties in that category instance, it only helps answer a question (belongs to or doesn't belong).

For ResourceModel, if we require that each instance of *vf:Resource* which references it with *vf:model* keeps copy of all the properties that apply to that instance. I would say again that it **does NOT** work like 'TypeObject' and seems similar to combination of category (very specific one) and template. I think we can continue exploring that in issue #179 

To emphasize it again, not of the instances seem to rely on properties from referenced category or template, for templates they don't even have to keep referencing them at all. This doesn't seem like 'TypeObject' pattern in action where to my understanding instances 'rely on' (delegate) some of the properties to their 'TypeObjects'.

**Comment by elf-pavlik** [comment]
> Action used to be EventType in NRP. It encapsulates the resource effect (increment, decrement, none, etc) and whether it applies to a process or an exchange (or maybe both?) Otherwise those behaviors get duplicated all over Events, Commitments, Recipes, etc.

Just to clarify something for everyone not familiar with NRP model. EventType from NRP maps to Action in VF, and NOT EventTemplate which we just renamed to IntentTemplate.

If we would say, Template == Type and what we called EventTemplate change to EventType, then nrp:EventType != vf:EventType.

> [Edit2] And yes, @elf-pavlik , I'm still using CommitmentType in NRP. It's a legacy. We'll see if the name survives the refactoring...I'm not pushing it here in VF, that table was a copy-pasta job.

To my understanding what we currently call IntentTemplate, matches what in that table appears as *nrp:CommitmentType*. While it seems that relation between *nrp:CommitmentType* and *nrp:EventType* doesn't clearly follow relation between *nrp:Commiement* and *nrp:Event*. BTW currently VF has *vf:Intent* and *vf:Commitment* while NRP uses *nrp:Commitment* for both (so even if no one actually committed to anything).

**Comment by fosterlynn** [comment]
> Instances of Intent (possibly also directly Commitment or Event) created based on IntentTemplate don't have to reference that tempate at all, so NOT a type object.

@elf-pavlik I have not suggested that IntentTemplate be anythingType, so don't worry please.

And in fact I don't care hugely how we name the ProcessType and ResourceType and ExchangeType (if it exists) classes (related to other issues), which is why I was ok with Template, until I found out that was actually a different thing to you.  Do you have a suggestion that can work for all?  It seemed like others were ok with XType.  I may define type object a bit looser than you - going back to the good ol' days of type files when ISAM was a revolution.  It has an essence that is recognizable, and has been around a long time.

And yes what Pavlik said in his last post about the various evolutions of naming is very true, sorry this is so confusing.

**Comment by fosterlynn** [comment]
>And yes what Pavlik said in his last post about the various evolutions of naming is very true, sorry this is so confusing.

NRP was a deep dive into a particular set of use cases, and very experimental.  VF is trying to take that experience and make it useful generally, melding it with the other experiences and skills and use cases of other people.  I am quite positive that VF will be much better than NRP.

**Comment by fosterlynn** [comment]
> instance of vf:Process created based on this template, doesn't have to reference it at all. 

 Actually, what I was proposing to call ProcessType, ResourceType, and ExchangeType (if it exists) all benefit from being maintained as a reference, so that you can see all the whatevers of that type.

**Comment by elf-pavlik** [comment]
> Actually, what I was proposing to call ProcessType, ResourceType, and ExchangeType (if it exists) all benefit from being maintained as a reference, so that you can see all the whatevers of that type.

In that case this 'type' plays similar role to 'category' - belongs to or not, but still doesn't act as 'TypeObject' (no delegation)

> There’s a lot of similarity between this pattern and the State pattern. Both patterns let an object delegate part of what defines itself to another object. With a type object, we’re usually delegating what the object is: invariant data that broadly describes the object. With State, we delegate what an object is right now: temporal data that describes an object’s current configuration. - http://gameprogrammingpatterns.com/type-object.html

If we talk about 'TypeObjects' I'd like to check for each case what instances would delegate to that 'TypeObject'. Only placement in some taxonomy graph (hierarchy with multiple parents) or something besides that. In case of category no other properties of category instance matter other then parent/child (broader/narrower) which position them in taxonomy.

**Comment by fosterlynn** [comment]
also from the game programming reference:

>Define a type object class and a typed object class. Each type object instance represents a different logical type. Each typed object stores a reference to the type object that describes its type.

>Instance-specific data is stored in the typed object instance, and data or behavior that should be shared across all instances of the same conceptual type is stored in the type object. Objects referencing the same type object will function as if they were the same type. This lets us share data and behavior across a set of similar objects, much like subclassing lets us do, but without having a fixed set of hard-coded subclasses.

>This pattern is useful anytime you need to define a variety of different “kinds” of things, but baking the kinds into your language’s type system is too rigid. In particular, it’s useful when either of these is true:
>* You don’t know what types you will need up front. (For example, what if our game needed to support downloading content that contained new breeds of monsters?)
>* You want to be able to modify or add new types without having to recompile or change code.

I think this describes pretty well what we are trying to do here.

> but still doesn't act as 'TypeObject' (no delegation)

ProcessType has a name, estimated duration, and notes (sometimes very detailed notes about what to do in the process).  Those can all be delegated.

ResourceType has name, substitutable behavior, unit, image, notes, which can all be delegated.

But I repeat, if there is another term that works for all of us, let's use it.

**Comment by bhaugen** [comment]
Process Types also often describe the instructions for executing a process of that type. No need for every process to repeat them. For example: https://en.wikipedia.org/wiki/List_of_manufacturing_processes

In a particular environment, those process types will have even more particular instructions.

**Comment by elf-pavlik** [comment]
> ResourceType has name, substitutable behavior, unit, image, notes, which can all be delegated.

We had conversation earlier about *unit* and that if ResourceType stayed responsible for it, types couldn't get shared between metric and imperial systems. Similar resources which sometimes get managed by kg, sometimes by liter and sometimes by piece (eg melon).

**Comment by bhaugen** [comment]
>  if ResourceType stayed responsible for it, types couldn't get shared between metric and imperial systems

Good issue. Got a suggested solution? We should probably look at how Gs1 does it...

> Similar resources which sometimes get managed by kg, sometimes by liter and sometimes by piece (eg melon).

Units are another deep topic. Operational systems usually need unit conversions, too: conversions between one unit and another. qudt gets into conversions between standard types, but real life also has resourcetype-specific units like case or package. Like 24 bottles in a case, 355 ml per bottle.

One question is how deep do we want to get into units in VF?

**Comment by bhaugen** [comment]
I don't want to belabor Type Objects too much anymore, since we are not actually using the term in VF, but one of the main reason that they emerge over and over again, in many settings, is that they are user-defined. So people can have as many as they want, and nobody needs much technical knowhow to create another.

That also brings with it the problem of agreement on types between communities, but as Decentrale has found, those are good conversations to have. That's part of how communities learn to work together. (As we are finding here in VF: it's not easy, but we learn how to do it, or we fail. And in the process, we learn more about each other and processes of collaboration.)

**Comment by elf-pavlik** [comment]
> I don't want to belabor Type Objects too much anymore, since we are not actually using the term in VF, but one of the main reason that they emerge over and over again, in many settings, is that they are user-defined. So people can have as many as they want, and nobody needs much technical knowhow to create another.

We had conversation with @fosterlynn that things like 'decrementing/incrementing' which you two often refer to as 'resource effects' can't stay fully 'user defined' since the code needs to handle that + / - logic and preferably have unit tests for it.

> ProcessType has a name, estimated duration, and notes (sometimes very detailed notes about what to do in the process). Those can all be delegated.
> Process Types also often describe the instructions for executing a process of that type. No need for every process to repeat them. For example: https://en.wikipedia.org/wiki/List_of_manufacturing_processes
> In a particular environment, those process types will have even more particular instructions.

I think whenever we talk about *delegation* we need to think about mutability/immutability. Let's say some observed process references something you call Process Type which has those 'detailed notes about what to do'. At some point in a future someone updates those notes since 'now we do it like this'. Old processes which got executed based on old notes, will now reference Process Type which notes doesn't match notes used during execution of that process.
It seems even more relevant in open web ecosystem then in closed silo, since if processes managed by agent A reference Process Types controlled by agent B, agent B may update those Process Types without coordinating in any way with agent A, or any other agent whose processes may reference those Process Types.

**Comment by bhaugen** [comment]
> We had conversation with @fosterlynn that things like 'decrementing/incrementing' which you two often refer to as 'resource effects' can't stay fully 'user defined' since the code needs to handle that + / - logic and preferably have unit tests for it.

We agree about resource effects, and our EventTypes in our NRP software are not actually user defined, they are system controlled. They started out being user defined, but as you suggest, that made a mess...

>  Let's say some observed process references something you call Process Type which has those 'detailed notes about what to do'. At some point in a future someone updates those notes since 'now we do it like this'. Old processes which got executed based on old notes, will now reference Process Type which notes doesn't match notes used during execution of that process.

I think ProcessTypes need to be versioned. That takes care of the mutability problem. Probably some other Types as well.

P.S. @elf-pavlik are you ok with ProcessType and ResourceType as names and concepts, or is this still one of those circular unending arguments?

**Comment by fosterlynn** [comment]
I actually don't that much care if they are type objects or are called XType, although it is interesting.  I do care that we preserve the ability for people to publish open source hardware designs that others can use.  (For one example.)  

Where this started was this requirement vs another requirement that people in a group can create templates they can use to further help with automating planning, and include all their group specific data elements, like specific people who are providers and receivers, etc.  I believe this came out of the Vientos work.

It seems there is some agreement to support both requirements.  We just have to name them appropriately (post suggestions if you don't like XType!) and define the templating, which I assume can be called XTemplate.  And there is still some work to be done on recipes, like features and options, and some questions I have about specific model type recipes and more general process flow type recipes.

I've been a little hot to trot on some of the naming because this stuff is actually being looked at by various people who are interested in using it, and it would be good if we could settle on a few things.  Of course, these things also do need to proceed at their own internal pace.  :)

**Comment by bhaugen** [comment]
> I've been a little hot to trot on some of the naming because this stuff is actually being looked at by various people who are interested in using it, and it would be good if we could settle on a few things. 

Plus, I want to start work on [the Vocabulator](https://github.com/valueflows/vocabulator), which I hope will make this VF stuff come to life a bit more. Although actual use in real life communities will be even livelier.

**Comment by elf-pavlik** [comment]
> I think ProcessTypes need to be versioned. That takes care of the mutability problem. Probably some other Types as well.

cooking salsa v1 != cooking salsa v2 and possibly identifiers (URIs) of those versions need to differ for the versions which gives us 2 different process types which has some relation. especially if different agents need to create versions independently so we can have many different v2 (practically forking)

> P.S. @elf-pavlik are you ok with ProcessType and ResourceType as names and concepts, or is this still one of those circular unending arguments?

I don't think we already clarified what exactly we try to name and how we mode it. As we see the delegation part doesn't has various gotchas.

> I actually don't that much care if they are type objects or are called XType, although it is interesting. I do care that we preserve the ability for people to publish open source hardware designs that others can use. (For one example.)

With open source hardware we very likely will have forks, so again we currently don't have strategy for versions and forks.

> Where this started was this requirement vs another requirement that people in a group can create templates they can use to further help with automating planning, and include all their group specific data elements, like specific people who are providers and receivers, etc.

Besides having something like abstract type/class (not necessarily 'object type') to use as a template. I also see possibility of using any instance as a template (more in direction of prototype). We had conversations earlier that one could create 'template/type' automatically based on instance of a Process. Possibly processes can get created based on other instance of a Process used as prototype for now process, where I would stay cautious with delegation and copy all relations on new process so it doesn't even have to reference the prototype process (just as I suggested process may not need to reference 'template' used to create it - so not exactly 'type object').

**Comment by bhaugen** [comment]
Maybe we should quit using the phrase 'type object' since it seems to be a boogeyman to @elf-pavlik - just always ends up with a lot of distraction, and we never make any progress.

And we obviously have not defined what we mean by "type" and "template" very well. 

One of the main uses of Type in economic relationships is to say definitively what something is. As in, Bike or Apple? Or Delicious Apple or Haralson Apple? While I like prototypes in general, I suspect they will not be as good for economic relationships as declared {whatever word you want here that means Type).

**Comment by elf-pavlik** [comment]
> One of the main uses of Type in economic relationships is to say definitively what something is. As in, Bike or Apple? Or Delicious Apple or Haralson Apple?

I think common Taxonomies (Wikidata, Agrovoc or even better all those interlinked) can take care of such categorization / classification. But I don't think those taxonomies should also stay responsible for providing recipes with 'detailed steps' - possibly two different concerns.

**Comment by bhaugen** [comment]
> I think common Taxonomies (Wikidata, Agrovoc or even better all those interlinked) can take care of such categorization / classification. 

Or Motobecane Fantom Bike or Gravity Bullseye Monster Bike?

>  I don't think those taxonomies should also stay responsible for providing recipes with 'detailed steps' - possibly two different concerns.

Agreed.

**Comment by elf-pavlik** [comment]
If we look at observed process based on snippet from @fosterlynn 

``` yaml
'@context': https://w3id.org/valueflows/v1-dev
'@graph':
  - '@id': ex:cooked-salsa-dfa09fa9fja
    '@type':
      - vf:Process
    'vf:duration': PT5H #https://en.wikipedia.org/wiki/ISO_8601#Durations
    'vf:note': 'Combine all ingredients, ....'
    'vf:input':
      - '@id': '#input-1'
        '@type': vf:Event
        'vf:action': vf:consume
        'vf:affects':
          '@id': ex:resource-2535252 
          'vf:category': prodont:Tomato
        'vf:affectedQuantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Kilogram
          'quid:numericValue': 4
      - '@id': '#input-2'
        '@type': vf:Event
        'vf:action': vf:consume
        'vf:affects':
          'vf:resource': ex:resource-242094242
          'vf:category': prodont:Onion
        'vf:quantity':
          '@type': quid:QuantityValue
          'quid:unit': quid:Kilogram
          'quid:numericValue': 2
      - '@id': '#process-input-4'
        '@type': vf:Event
        'vf:action': vf:work
        'vf:category': ex:Cooking
        'vf:provider': ex:Bob
        'vf:quantity':
            '@type': quid:QuantityValue
            'quid:unit': quid:Hour
            'quid:numericValue': 3
    'vf:output':
      - '@id': '#output-1'
        '@type': vf:Event
        'vf:action': vf:produce
        'vf:resource':
          '@type': vf:Resource
          '@id': ex:resource-34224424
          'vf:model': ex:our-cool-salsa
        'vf:quantity':
          '@type': quid:QuantityValue
          'quid:unit': ???:Each #or however we do this
          'quid:numericValue': 12
```

Creating new Plan of a Process based on the instance of observed Process, seems comparable to creating new Plan of a Process based on that 'abstracted' Recipe with 'abstract' Type/Template classes. ATM I don't say that one way should work better then the other one, just that we don't have 'obvious' way do do it here and we should patiently evaluate pros and cons of various approaches. Keeping in mind that we don't design for closed silo system but for the open web ecosystem.

**Comment by elf-pavlik** [comment]
>> @elf-pavlik:  I don't think those taxonomies should also stay responsible for providing recipes with 'detailed steps' - possibly two different concerns.

> @bhaugen: Agreed.

Maybe we can try with ResourceTaxonomy, ProcessTaxonomy, and ExchangeTaxonomy for the taxonomy layer (or more column). Where we don't expect it to stay responsible for Recipes, but only position in taxonomy graph (multi parent) - answering question *belongs to a particular (sub)set of things or doesn't belong to it*. Pretty much *vf:category*

Recipes we can investigate in parallel and see how they connect to those taxonomies, but without expecting taxonomies to address concern of recipes.

**Comment by bhaugen** [comment]
If you want to get down and dirty with recipes, I am for it. Maybe start another  issue + thread?

**Comment by fosterlynn** [comment]
>Maybe we can try with ResourceTaxonomy, ProcessTaxonomy, and ExchangeTaxonomy for the taxonomy layer (or more column). 

Good point, taxonomy / category could be applicable to all of these.

>Recipes we can investigate in parallel and see how they connect to those taxonomies, but once again without expecting taxonomies to address concern of recipes.

Yes, fits with some things I've been mulling over when trying to integrate exchange into recipes.  Let's give it a try.

**Comment by fosterlynn** [comment]
>Creating new Plan of a Process based on the instance of observed Process, seems comparable to creating new Plan of a Process based on that 'abstracted' Recipe with 'abstract' Type/Template classes.

Yes, seems like another good way to do it.

**Comment by bhaugen** [comment]
> Creating new Plan of a Process based on the instance of observed Process, seems comparable to creating new Plan of a Process based on that 'abstracted' Recipe with 'abstract' Type/Template classes.

Sensorica wanted to do that, when they were into manufacturing products, but what they wanted was to take a prototype process flow and turn it into a recipe, so they could also edit it as a recipe.

Another consideration if you just went for a plan with no recipe would be when your plan traipsed over into some resource type that had a recipe. I don't know that it would be a problem, but would want to do a bunch of real-world testing before I deployed the code...

**Comment by elf-pavlik** [comment]
I just made PR to capture what seems resolution on keeping Taxonomies distinct from Recipes.

> Sensorica wanted to do that, when they were into manufacturing products, but what they wanted was to take a prototype process flow and turn it into a recipe, so they could also edit it as a recipe.

They could also edit it as plan which would later lead to Observation that could serve as a base for another Plan. Possibly one could also base new Plan on old Plan even if it never got to the point of Observation. I guess we may discuss all those options in dedicated issue and close this one after we merge that Taxonomies related PR?

**Comment by fosterlynn** [comment]
Resolved for v.3, please re-open if needed.
